### ISSUE TITLE
Redis cache TTL

### ISSUE THREAD

**Issue by HitkoDev** [apache/incubator-pagespeed-mod#1704]
I've had a site with PageSpeed and Redis cache running for a few months now, and when running `redis-cli INFO` I've noticed this:

```
db1:keys=840116,expires=0,avg_ttl=0
```

It seems like PageSpeed doesn't set expire for the shared memory cache, keeping lots of unnecessary keys in the cache. Is there a way to set the expire time?

**Comment by oschaaf** [comment]
As far as I can tell, there is not a way for you to set the expire time.  
Actually, I think that you are observing the intended behavior. The system relies on (eviction) policies of the physical cache backends (like Redis) to enforce constraints like max cache size and max number of entries.

**Comment by HitkoDev** [comment]
Then perhaps we should consider adding the option to set expiration? Having a designated Redis instance for each site creates a lot of unnecessary overhead, and by specifying TTL, sites with different caching needs could use the same Redis instance. 

Another use case for TTL is to keep metadata for low access rate content cached, while refreshing metadata for dynamic content on a regular basis.

**Comment by oschaaf** [comment]
You can configure PageSpeed to use distinct Redis databases per site [1], which I hope avoids the need for separate Redis instances for this purpose? As far as I can tell, it is also possible to configure size constraints and eviction policies in Redis per database, allowing you to express what is appropriate for a site, which covers your use-case above? 

Having said that, technically I suspect it will not be not hard to accomplish what you suggest, but I still wonder if there's a clear benefit in doing so.. 

[1] https://www.modpagespeed.com/doc/system#redis

**Comment by HitkoDev** [comment]
I'm well aware of that; why would I want to mess with PageSpeed cache config, if I was forced to use separate Redis instances anyway? 

So far I was unable to set Redis config per database, and all resources I've came across suggest that's not possible. Could you provide some instructions, if you say it's possible?

**Comment by oschaaf** [comment]
I've never touched Redis, so that's going to be hard. But googling for "redis database evicition polciy", the  following looks promising:
https://redislabs.com/redis-enterprise-documentation/administering/database-operations/eviction-policy/
And below that:
https://redislabs.com/redis-enterprise-documentation/administering/database-operations/memory-limit/

**Comment by HitkoDev** [comment]
The provided links are for Redis Enterprise cloud service, where the term "database" refers to a (configurable) collection within the customer's cluster. With on-prem Redis (as well as AWS ElastiCache and most other cloud-based Redis services), database is a collection within a single Redis instance, and the [documentation](https://redis.io/topics/config) only mentions configuring an entire Redis instance.

**Comment by oschaaf** [comment]
Ah I see, thanks for clarifying. I agree it would be nice to be able to experiment with setting expiry on Redis keys to see how that works out. I don't think I'm going to be able to work on such a feature for a while, but if you want you can take a stab at it I would be more then willing to help out getting it done. A recent PR that is in the same space was:
https://github.com/pagespeed/mod_pagespeed/pull/1627/files

That PR adds an option, and uses that option to perform redis database selection. It may be a good starting point, if you want to have a go at it :)

**Comment by HitkoDev** [comment]
I'll give it a try, seems like it should be pretty simple to get it working.

**Comment by oschaaf** [comment]
Great! Let me know if you need any help!

**Comment by oschaaf** [comment]
@HitkoDev I was curious, any updates on this?

**Comment by HitkoDev** [comment]
Unfortunately I didn't manage to find time for this yet, because I have some projects with higher priority.

**Comment by danDanV1** [comment]
Essentially, all that should be required is swapping the SET command with SETEX, and defining a default for the TTL to -1 (no expiry), which will match the current pagespeed behaviour. Then defining a ModPagespeedRedisTtl in config.

```SETEX keyname ttltime value```
https://redis.io/commands/setex

/pagespeed/system/redis_cache.cc
```
void RedisCache::Put(const GoogleString& key, const SharedString& value) {
  RedisReply reply = RedisCommand(
      LookupConnection(key),
      "SET %b %b",
      {REDIS_REPLY_STATUS},
```

I'd submit a pull, but have no experience in c++

**Comment by jmarantz** [comment]
Agreed this shouldn't be difficult, but adding a new config and plumbing it through to this function does touch probably ~10 files; various RewriteOptions* classes, cache-configuration code, etc.

This would need to be a ServerScope option, and would need to be incorporated in to a key that is used in a map to share cache-instances between multiple similarly configured servers. This is straightforward but does involve touching a bunch of code to plumb the option through.

**Comment by jmarantz** [comment]
RE the original premise of this feature-request; I have a couple of questions:

1. Won't Redis evict uninteresting keys based on some reasonable policy (LFU, LRU)?
2. What would you set the TTL to, globally?  A year?

**Comment by danDanV1** [comment]
I would suggest a default -1 expiry to match current pagespeed config behaviour, and yes we'd probably use a long TTL like 1 year in our config.

The default LRU for Redis and also most desirable for our use case is `volatile-lru`: evict keys by trying to remove the less recently used (LRU) keys first, but only among keys that have an expire set, in order to make space for the new data added.

Without a TTL set from pagespeed, the pagespeed cached keys will never be purged using a volatile-lru policy.

To prevent running out of cache space with pagespeed, redis imlementations are forced to use `allkeys-lru`: evict keys by trying to remove the less recently used (LRU) keys first, in order to make space for the new data added.

`allkeys-lru` policy is a problem if you are sharing the redis cache with other applications and there are keys that you wish to persist as they may be purged before some ephemeral keys.

By defining a ttl and using `volatile-lru` you can explicitly declare which keys you want to evict as the cache fills up.

**Comment by stufan** [comment]
We need this feature in place as well. For now we solve it deploying pagespeed cache on a dedicated redis server.

**Comment by oschaaf** [comment]
The feature is ready, but needs test coverage. Re-opening
